### PR TITLE
expression: support push down right/left to TiKV

### PR DESCRIPTION
**Pull request by Yisaer**
### What problem does this PR solve?

Issue Number: ref https://github.com/pingcap/tidb/issues/31846

### What is changed and how it works?

support push down right/left

### Check List

Tests 

- [x] Unit test

### Release note

```release-note
None
```

### CODE REVIEW

**Review by rebelice — COMMENTED** [review]
Please add integration testing for this PR

**Comment by Yisaer** [comment]
> Please add integration testing for this PR

updated.

File: planner/core/integration_test.go [hunk 1]
```diff
@@ -2479,6 +2479,25 @@ func (s *testIntegrationSerialSuite) TestIssue18984(c *C) {
...
+	}
+	tk.MustQuery("explain analyze select /*+read_from_storage(tikv[t])*/ * from t where right(c,1);").
+		CheckAt([]int{0, 3, 6}, rows)
+	rows[1][2] = "left(test.t.c, 1)"
+	tk.MustQuery("explain analyze select /*+read_from_storage(tikv[t])*/ * from t where left(c,1);").
+		CheckAt([]int{0, 3, 6}, rows)
+	rows[1][2] = "mod(test.t.id, 1)"
+	tk.MustQuery("explain analyze select /*+read_from_storage(tikv[t])*/ * from t where mod(id,1);").
+		CheckAt([]int{0, 3, 6}, rows)
+}
```
Comment (rebelice) [inline planner/core/integration_test.go:2499]: Please insert some data and check the result.

File: planner/core/integration_test.go [hunk 2]
```diff
@@ -2479,6 +2479,24 @@ func (s *testIntegrationSerialSuite) TestIssue18984(c *C) {
...
+		{"TableReader_7", "root", "data:Selection_6"},
+		{"└─Selection_6", "cop[tikv]", "right(test.t.c, 1)"},
+		{"  └─TableFullScan_5", "cop[tikv]", "keep order:false, stats:pseudo"},
+	}
+	tk.MustQuery("explain analyze select /*+read_from_storage(tikv[t])*/ * from t where right(c,1);").
+		CheckAt([]int{0, 3, 6}, rows)
+	rows[1][2] = "left(test.t.c, 1)"
+	tk.MustQuery("explain analyze select /*+read_from_storage(tikv[t])*/ * from t where left(c,1);").
+		CheckAt([]int{0, 3, 6}, rows)
+	rows[1][2] = "mod(test.t.id, 1)"
```
Comment (rebelice) [inline planner/core/integration_test.go:2497]: This line should be removed.

**Review by rebelice — COMMENTED** [review]
rest LGTM

**Review by LittleFall — APPROVED** [review]
this pr LGTM, by the way, are there some links showing that the implementation of these two functions in tikv is well tested?

**Comment by Yisaer** [comment]
> this pr LGTM, by the way, are there some links showing that the implementation of these two functions in tikv is well tested?

In https://github.com/tikv/tikv/issues/5751, it seems `left` and `right` are well handled now.

**Review by rebelice — APPROVED** [review]

**Pull request merged**